src/main/java/org/openstreetmap/atlas/checks/validation/linear/edges/InvalidPiersCheck.java

```diff
@@ -0,0 +1,350 @@
+     * @param originalEdge
+     *            originalEdge
+     * @return true if the input edges form a polygon geometry,
+     */
+    private boolean hasPolygonalGeometry(final List<Edge> edgesFormingOSMWay,
+            final Edge originalEdge)
+    {
+        // Captures cases where the single edge has polygonal geometry
+        if (edgesFormingOSMWay.size() == 1 && originalEdge.start().equals(originalEdge.end())
+                && originalEdge.length().isGreaterThan(Distance.ZERO))
+        {
+            return true;
+        }
+        final HashSet<Long> wayIds = new HashSet<>();
+        Edge nextEdge = originalEdge;
+        // Loop through out going edges with the same OSM id
+        while (nextEdge != null)
+        {
+            wayIds.add(nextEdge.getIdentifier());
+            final List<Edge> nextEdgeList = Iterables.stream(nextEdge.outEdges())
+                    .filter(outEdge -> outEdge.isMasterEdge()
+                            && outEdge.getOsmIdentifier() == originalEdge.getOsmIdentifier())
+                    .collectToList();
+            nextEdge = nextEdgeList.isEmpty() ? null : nextEdgeList.get(0);
+            // If original edge is found, the way is closed
+            if (nextEdge != null && wayIds.contains(nextEdge.getIdentifier()))
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Checks if given edge is connected to a ferry/building or overlaps a building. In case of
+     * polygonal piers, buildings that are within the polygon and connected to it are considered
+     * overlapping buildings. In case of linear piers, we only check if the building intersects any
+     * of the edges of the linear pier.
+     *
+     * @param originalEdge
+     *            any edge
+     * @param edgesOfOSMWay
+     *            list of edges forming OSM way
+     * @param isPolygonal
+     *            if true, the OSM way has polygonal geometry else a linear geometry
+     * @param polygon
+     *            polygon formed from the OSM way edges
+     * @return true if the edge is connected to a building/ferry or overlaps a building
+     */
+    private boolean isConnectedToFerryOrBuilding(final Edge originalEdge,
+            final List<Edge> edgesOfOSMWay, final boolean isPolygonal, final Polygon polygon)
+    {
+        final boolean intersectsFerryRoute;
+        final boolean intersectsBuilding;
+        final Atlas atlas = originalEdge.getAtlas();
+        // Checks if the pier is connected to a ferry route
+        intersectsFerryRoute = edgesOfOSMWay.stream()
+                .anyMatch(eachEdge -> eachEdge.connectedEdges().stream().filter(
+                        connectedEdge -> this.areOnSameLevelOrLayer(connectedEdge, originalEdge))
+                        .anyMatch(RouteTag::isFerry));
+        if (isPolygonal)
+        {
+            intersectsBuilding = edgesOfOSMWay.stream().anyMatch(eachEdge -> Iterables
```

Comment:
Couldn't we do this once for `polygon` instead of performing the check on each in Edge in `edgesOfOSMWay`.

Reply:
I had used polygon first like ```Iterables.stream(atlas.areasIntersecting(polygon,intersectingArea ->
                    this.areOnSameLevelOrLayer(intersectingArea, originalEdge)
                    && (IS_BUILDING.test(intersectingArea)
                    || IS_FERRY_TERMINAL.test(intersectingArea)))).iterator().hasNext()``` which gave me two FPs in a country. Adjacent areas where pulled in using the polygon. So I went for each edges.

Reply:
Interesting, I wounder if that has to do with polygonal vs liner intersection. It does seem like this is an expensive way to perform this calculation, as opposed to running the query for a single geometry. Do you think it would work to use linear intersection by querying/testing on a PolyLine made from all the way Edges?

Reply:
After some separate discussions it appears that there is a potential bug in area intersection. This seems like an ok solution while that is investigated further.